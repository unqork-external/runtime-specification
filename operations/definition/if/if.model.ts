import { DiscriminatorValue, Required } from '@tsed/schema'

import { IfOperationOptions } from './if.options'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DiscriminatorValue(OperationTypes.IF)
export class IfOperation extends Operation<OperationTypes.IF> {
  @Required()
  public options: IfOperationOptions = new IfOperationOptions()
}
