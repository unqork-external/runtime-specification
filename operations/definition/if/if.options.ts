import { Any, CollectionOf, Description, Example, Optional, Required } from '@tsed/schema'

import { TrimmedDescription } from '../../../../decorators/schema/trimmedDescription.decorator'
import { Operation, OperationOptions } from '../../interface'

export class Condition {
  @Required()
  @Description('The condition operation to run')
  operation: Operation

  @Optional()
  @Example('isValid')
  @Description(`
    resultProperty refers to the property to expect the result from
    If omitted, resultValue is compared against entire result
  `)
  resultProperty?: string

  @Any()
  @Required()
  @Example(true)
  @Description('resultValue refers to the expected value.')
  resultValue: unknown
}

export class IfOperationOptions extends OperationOptions {
  @CollectionOf(Condition)
  @TrimmedDescription(`
    The set of conditions to fire or evaluate.
    This could be represented as something as simple as firstName.value == "john" or
    an operation like VALIDATE.
  `)
  conditions: Condition[]

  @CollectionOf(Operation)
  @TrimmedDescription(`
    Optionally set a bunch of operations to fire when if the conditions return as successful.
    This is not intended to always be used - this is a convenience field in case it reads better for the user
    Operations continue firing if the conditions are true anyway. if you put all happy path execution in a \`whenTrue\`
    you risk causing a Pyramid of Doom antipattern for logical representation.
  `)
  whenTrue?: Operation[]

  @CollectionOf(Operation)
  @TrimmedDescription(`
    Optionally set a bunch of operations to fire when if the conditions return as unsuccessful.
    If unsuccessful, we'll run the false conditions, and continue down the operation chain.
  `)
  whenFalse?: Operation[]

  @CollectionOf(Operation)
  @TrimmedDescription(`
    Optionally set a bunch of operations to fire when if the conditions return as a failure.
    If the conditions fail and error, we'll run the fail operations, and terminate the operation chain.
  `)
  onFail?: Operation[]
}
