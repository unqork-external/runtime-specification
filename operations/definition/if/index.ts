export * from './if.model'
export * from './if.options'
