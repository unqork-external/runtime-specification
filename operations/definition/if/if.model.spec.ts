import { IfOperation } from './if.model'
import { generateSchemaAndValidate } from '../../../../utilities'

describe('Operation Model: If Tests', () => {
  it('test that the schema matches our expectations', () => {
    const response = generateSchemaAndValidate(IfOperation)
    const def = response.schema
    const d = {
      properties: def.properties,
      required: def.required,
      type: def.type,
    }
    const options = def.definitions.IfOperationOptions

    expect(options).toMatchSnapshot()

    expect(d).toMatchSnapshot()
  })
})
