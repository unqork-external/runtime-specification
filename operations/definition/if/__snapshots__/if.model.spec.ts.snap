// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operation Model: If Tests test that the schema matches our expectations 1`] = `
{
  "properties": {
    "conditions": {
      "description": "The set of conditions to fire or evaluate. This could be represented as something as simple as firstName.value == "john" or an operation like VALIDATE.",
      "items": {
        "$ref": "#/definitions/Condition",
      },
      "type": "array",
    },
    "inputs": {
      "description": "
    Inputs to be used within formulas in other operation options
  ",
      "items": {
        "$ref": "#/definitions/InputRef",
      },
      "type": "array",
    },
    "onFail": {
      "description": "Optionally set a bunch of operations to fire when if the conditions return as a failure. If the conditions fail and error, we'll run the fail operations, and terminate the operation chain.",
      "items": {
        "$ref": "#/definitions/Operation",
      },
      "type": "array",
    },
    "shouldExecute": {
      "description": "
    Conditionally execute this operation based on a formula.
    Can set to false to temporarily disabled when debugging
  ",
      "type": [
        "string",
        "boolean",
      ],
    },
    "whenFalse": {
      "description": "Optionally set a bunch of operations to fire when if the conditions return as unsuccessful. If unsuccessful, we'll run the false conditions, and continue down the operation chain.",
      "items": {
        "$ref": "#/definitions/Operation",
      },
      "type": "array",
    },
    "whenTrue": {
      "description": "Optionally set a bunch of operations to fire when if the conditions return as successful. This is not intended to always be used - this is a convenience field in case it reads better for the user Operations continue firing if the conditions are true anyway. if you put all happy path execution in a \`whenTrue\` you risk causing a Pyramid of Doom antipattern for logical representation.",
      "items": {
        "$ref": "#/definitions/Operation",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`Operation Model: If Tests test that the schema matches our expectations 2`] = `
{
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/IfOperationOptions",
    },
    "type": {
      "const": "IF",
      "examples": [
        "IF",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
